models/cnn/cnn.py

```diff
@@ -0,0 +1,167 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class CNN(nn.Module):
+    
+    def __init__(self, pretrain, init_weights=False, num_scales=1):
+        super(CNN, self).__init__()
+
+        self.pretrain = pretrain
+        # architecture
+        self.init_layer = self.make_layers(in_channels=1, out_channels=64, n_layers=2, maxpool_size=None, first=True)
+        self.layer1 = self.make_layers(in_channels=64, out_channels=128, n_layers=2, maxpool_size=5)
+        self.layer2 = self.make_layers(in_channels=128, out_channels=192, n_layers=3, maxpool_size=5)
+        self.layer3 = self.make_layers(in_channels=192, out_channels=256, n_layers=3, maxpool_size=5)
+        self.layer4 = self.make_layers(in_channels=256, out_channels=256, n_layers=3, maxpool_size=5)
+            
+        self.fp_dim = 128
+        self.num_scales = num_scales
+        self.conv_c5 = nn.Conv1d(256, self.fp_dim, 1, 1, 0)
+
+        if self.num_scales > 1:
+            self.conv_c4 = nn.Conv1d(256, self.fp_dim, 1, 1, 0)
+        
+        if self.num_scales > 2:
+            self.conv_c3 = nn.Conv1d(192, self.fp_dim, 1, 1, 0)
+        
+        if self.init_weights:
+            self._initialize_weights()
+
+    def _initialize_weights(self):
+        for m in self.modules():
+            if isinstance(m, nn.Conv1d):
+                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
+                if m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+            elif isinstance(m, nn.BatchNorm1d):
+                nn.init.constant_(m.weight, 1)
+                nn.init.constant_(m.bias, 0)
+
+    def make_layers(self, in_channels, out_channels, n_layers, maxpool_size, first=False):
+        layers = []
+        layers = layers + [MaxPool1d(maxpool_size)] if not first else layers
+
+        for i in range(n_layers):
+            conv1d = nn.Conv1d(in_channels, out_channels, kernel_size=3, padding=1)
+            layers += [conv1d, nn.BatchNorm1d(out_channels)]
+            if i == n_layers - 1:
+                layers += [ChannelGate(in_channels)]
+            layers += [nn.PReLU()]
+            in_channels = out_channels
+
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
```

Comment:
Please add type hints  and docstring/comment what this method outputs/does depending on the configuration (ex. how the self.pretrain flag is set)